StringImpl refcounting should be atomic

geoffreygaren opened the pull request:
#### 5caae6b082d7b5b47e6250c3b8189869abcc4574
<pre>
StringImpl refcounting should be atomic
<a href="https://bugs.webkit.org/show_bug.cgi?id=289962">https://bugs.webkit.org/show_bug.cgi?id=289962</a>
<a href="https://rdar.apple.com/147313174">rdar://147313174</a>

Reviewed by Antti Koivisto, Yusuke Suzuki, Chris Dumez, and Keith Miller.

This patch takes the first step to making WTF::String thread-safe: Making
m_refcount increment and decrement atomic. (Hash, flags, and the AtomString
table are not yet atomic. I plan to address those in my next patch.)

The direct code change is pretty small, but it measured as a performance
regression. So most of this code is optimization. I got rid of about 1/3 of
WTF::StringImpl refcounts on Speedometer and about 1/2 on JetStream.

ARM has a special ldadd instruction for fast atomics. Net net, it measures as
about 1.5X slower than just add. That&apos;s pretty good! But it still required
some optimizations to be affordable. Intel atomics are much slower, so it took
far more optimizations to get Intel across the finish line.

A/B testers suggest that we might still have a &lt; 0.5% JetStream regression on
Intel. I checked in with Yusuke, Keith, and Maciej, and we can move forward
because we have a large total JetStream speedup on Intel this year, so customers
on older hardware will still see a net speedup when they get new software.

* Source/JavaScriptCore/bytecode/DirectEvalCodeCache.cpp:
(JSC::DirectEvalCodeCache::setSlow):
* Source/JavaScriptCore/bytecode/DirectEvalCodeCache.h:
(JSC::DirectEvalCodeCache::CacheKey::CacheKey):
(JSC::DirectEvalCodeCache::CacheKey::hash const):
(JSC::DirectEvalCodeCache::CacheKey::operator== const):
(JSC::DirectEvalCodeCache::CacheLookupKey::CacheLookupKey):
(JSC::DirectEvalCodeCache::CacheLookupKey::hash const):
(JSC::DirectEvalCodeCache::CacheLookupKey::operator== const):
(JSC::DirectEvalCodeCache::CacheLookupKey::operator CacheKey const):
(JSC::DirectEvalCodeCache::CacheLookupKeyHashTranslator::hash):
(JSC::DirectEvalCodeCache::CacheLookupKeyHashTranslator::equal):
(JSC::DirectEvalCodeCache::get):
(JSC::DirectEvalCodeCache::set):
(JSC::DirectEvalCodeCache::tryGet): Deleted.
* Source/JavaScriptCore/interpreter/Interpreter.cpp:
(JSC::directEvalCacheKey):
(JSC::eval): Added a fast path for looking up a rope string without resolving
the rope when the right hand side of the rope is trivial.

* Source/JavaScriptCore/dfg/DFGOperations.cpp:
(JSC::DFG::JSC_DEFINE_JIT_OPERATION): No need to ref StringImpls here because
the Structure they came from ensures their lifetime in this scope.

* Source/JavaScriptCore/jit/JITOperations.cpp:
(JSC::putByVal):
(JSC::putByValOptimize):
(JSC::getByVal):
(JSC::JSC_DEFINE_JIT_OPERATION):
(JSC::getByValWithThis):
(JSC::getByValMegamorphic): Refactored fast paths not to ref strings.

* Source/JavaScriptCore/jit/JITWorklist.cpp:
(JSC::JITWorklist::enqueue): Dan and I collided here on
<a href="https://commits.webkit.org/293052@main">https://commits.webkit.org/293052@main</a>, but this version measured as a speedup
so I&apos;m keeping the slight change in how we load this value in case it matters.

* Source/JavaScriptCore/parser/Parser.cpp:
(JSC::Parser&lt;LexerType&gt;::parseInner):
(JSC::Parser&lt;LexerType&gt;::parseModuleSourceElements):
* Source/JavaScriptCore/parser/Parser.h:
(JSC::Scope::hasDeclaredVariable):
(JSC::Scope::hasLexicallyDeclaredVariable const):
(JSC::Scope::hasDeclaredParameter):
* Source/JavaScriptCore/parser/VariableEnvironment.cpp:
(JSC::VariableEnvironment::markVariableAsCapturedIfDefined):
(JSC::VariableEnvironment::markVariableAsCaptured):
(JSC::VariableEnvironment::markVariableAsImported):
(JSC::VariableEnvironment::markVariableAsExported):
* Source/JavaScriptCore/parser/VariableEnvironment.h:
(JSC::VariableEnvironment::contains const):
(JSC::VariableEnvironment::remove):
(JSC::VariableEnvironment::find):
(JSC::VariableEnvironment::find const):
* Source/JavaScriptCore/runtime/ArgList.h:
(JSC::MarkedVector::begin):
(JSC::MarkedVector::end): Removed some unnecessary conversions from StringImpl*
to RefPtr&lt;StringImpl&gt; for lookup.

* Source/JavaScriptCore/runtime/ArrayPrototype.cpp:
(JSC::fastJoin):
(JSC::JSC_DEFINE_HOST_FUNCTION): Added a fast path for joins that are all
strings.

* Source/JavaScriptCore/runtime/CacheableIdentifier.h:
* Source/JavaScriptCore/runtime/CacheableIdentifierInlines.h: Added a fast path
for grabbing a cacheable identifier without first asking if it exists.

* Source/JavaScriptCore/runtime/JSGlobalObjectFunctions.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION): No need for refcount churn if our StringImpl*
is owned by our structure.

* Source/JavaScriptCore/runtime/JSONAtomStringCache.h:
(JSC::JSONAtomStringCache::makeIdentifier): Deleted.
* Source/JavaScriptCore/runtime/JSONAtomStringCacheInlines.h:
(JSC::JSONAtomStringCache::makeIdentifier):
(JSC::JSONAtomStringCache::existingIdentifier):
(JSC::JSONAtomStringCache::make): Deleted. Added a fast path for grabbing a
StringImpl* without refcount churn.

* Source/JavaScriptCore/runtime/JSObject.cpp:
(JSC::JSObject::putOwnDataPropertyBatching):
(JSC::JSObject::putDirectForJSONSlow): Deleted.
* Source/JavaScriptCore/runtime/JSObject.h:
* Source/JavaScriptCore/runtime/JSObjectInlines.h:
(JSC::JSObject::putDirectInternal): Our fast path doesn&apos;t deterministically rule
out a hit anymore, so we can&apos;t skip that check on the slow path anymore.

* Source/JavaScriptCore/runtime/JSString.cpp:
(JSC:: const):
(JSC::JSRopeString::resolveRopeToAtomString const): Deleted.
(JSC::JSRopeString::resolveRopeToExistingAtomString const): Deleted.
* Source/JavaScriptCore/runtime/JSString.h:
(JSC::JSString::getValueImpl const):
(JSC::JSString::tryGetValueImpl const):
(JSC:: const):
(JSC::JSString::toAtomString const): Deleted.
(JSC::JSString::toExistingAtomString const): Deleted.
* Source/JavaScriptCore/runtime/JSStringInlines.h:
(JSC::JSRopeString::tryGetLHS const): Use GCOwnedDataScope for more return values, to avoid
refcount churn.

* Source/JavaScriptCore/runtime/JSStringJoiner.cpp:
(JSC::joinStrings):
(JSC::JSOnlyStringsJoiner::joinSlow):
* Source/JavaScriptCore/runtime/JSStringJoiner.h:
(JSC::JSOnlyStringsJoiner::JSOnlyStringsJoiner):
(JSC::JSOnlyStringsJoiner::tryJoin): Helper class for the fast path for joining
only strings.

* Source/JavaScriptCore/runtime/LiteralParser.cpp:
(JSC::reviverMode&gt;::equalIdentifier):
(JSC::reviverMode&gt;::existingIdentifier):
(JSC::reviverMode&gt;::makeJSString):
(JSC::requires):
* Source/JavaScriptCore/runtime/LiteralParser.h: Added a fast path for following
a property transition without looking up and/or ref&apos;ing an Identifier.

* Source/JavaScriptCore/runtime/NumericStrings.h: Made the cache bigger because
it made a ~20% difference on hit rate, reducing string churn.

* Source/JavaScriptCore/runtime/ObjectConstructor.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/JavaScriptCore/runtime/ObjectConstructorInlines.h:
(JSC::objectAssignFast): No need to ref strings owned by our structure.

* Source/JavaScriptCore/runtime/StringPrototype.cpp:
(JSC::substituteBackreferences):
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/JavaScriptCore/runtime/StringPrototypeInlines.h:
(JSC::jsSpliceSubstringsWithSeparator):
(JSC::jsSpliceSubstrings):
(JSC::removeAllUsingRegExpSearch):
(JSC::addToRegExpSearchCache):
(JSC::replaceAllWithCacheUsingRegExpSearchThreeArguments):
(JSC::replaceAllWithCacheUsingRegExpSearch):
(JSC::replaceAllWithStringUsingRegExpSearchNoBackreferences):
(JSC::replaceAllWithStringUsingRegExpSearch):
(JSC::replaceOneWithStringUsingRegExpSearch):
(JSC::replaceUsingRegExpSearch):
(JSC::removeUsingRegExpSearch): Deleted.
(JSC::replaceUsingRegExpSearchWithCache): Deleted.
* Source/JavaScriptCore/runtime/Structure.cpp:
(JSC::StructureTransitionTable::add): Streamlined the fast paths and added
a new fast path for a single separator join.

* Source/WTF/wtf/ValueOrReference.h: Made a new type to encapsulate a pattern
where we usually return an existing reference, but sometimes need to allocate
a temporary. This avoids a copy in the case where we return an existing reference.
Mostly deployed this in form controls code.

* Source/WTF/wtf/Vector.h:
(WTF::insertInUniquedSortedVector): Factored out a helper function for use in
HTML attributes.

* Source/WTF/wtf/text/StringImpl.cpp:
* Source/WTF/wtf/text/StringImpl.h:
(WTF::StringImpl::ref):
(WTF::StringImpl::deref): Aha! The actual refcounting change. (Note that memory order
is relaxed because we have no data dependencies yet. That will come next.)

* Source/WebCore/dom/Attribute.h:
(WebCore::Attribute::Attribute): Don&apos;t copy strings when moving.

* Source/WebCore/dom/CharacterData.h: Don&apos;t copy the empty string unless you&apos;re
actually going to use it.

* Source/WebCore/dom/ElementTextDirection.cpp:
(WebCore::computeAutoDirectionality):
* Source/WebCore/dom/Text.cpp:
(WebCore::Text::create): Deleted.
* Source/WebCore/dom/Text.h:
(WebCore::Text::create):
* Source/WebCore/editing/cocoa/AutofillElements.cpp:
(WebCore::AutofillElements::computeAutofillElements):
* Source/WebCore/editing/cocoa/HTMLConverter.mm:
(HTMLConverter::_processElement):
* Source/WebCore/html/BaseCheckableInputType.cpp:
(WebCore::BaseCheckableInputType::fallbackValue const):
* Source/WebCore/html/BaseCheckableInputType.h:
* Source/WebCore/html/BaseDateAndTimeInputType.cpp:
(WebCore::BaseDateAndTimeInputType::hasBadInput const):
(WebCore::BaseDateAndTimeInputType::sanitizeValue const):
(WebCore::BaseDateAndTimeInputType::updateInnerTextValue):
(WebCore::BaseDateAndTimeInputType::setupDateTimeChooserParameters):
* Source/WebCore/html/BaseDateAndTimeInputType.h:
* Source/WebCore/html/ColorInputType.cpp:
(WebCore::ColorInputType::fallbackValue const):
(WebCore::ColorInputType::sanitizeValue const):
(WebCore::ColorInputType::valueAsColor const):
* Source/WebCore/html/ColorInputType.h:
* Source/WebCore/html/DateTimeLocalInputType.cpp:
(WebCore::DateTimeLocalInputType::sanitizeValue const):
* Source/WebCore/html/DateTimeLocalInputType.h:
* Source/WebCore/html/EmailInputType.cpp:
(WebCore::EmailInputType::sanitizeValue const):
* Source/WebCore/html/EmailInputType.h:
* Source/WebCore/html/HTMLInputElement.cpp:
(WebCore::HTMLInputElement::tooShort const):
(WebCore::HTMLInputElement::tooLong const):
(WebCore::HTMLInputElement::computeValidity const):
(WebCore::HTMLInputElement::listOptionValueAsDouble):
(WebCore::HTMLInputElement::value const):
(WebCore::HTMLInputElement::valueWithDefault const):
(WebCore::HTMLInputElement::setValue):
(WebCore::HTMLInputElement::sanitizeValue const):
* Source/WebCore/html/HTMLInputElement.h:
* Source/WebCore/html/HTMLTextAreaElement.cpp:
(WebCore::HTMLTextAreaElement::childrenChanged):
(WebCore::HTMLTextAreaElement::appendFormData):
(WebCore::HTMLTextAreaElement::value const):
(WebCore::HTMLTextAreaElement::setValueCommon):
(WebCore::HTMLTextAreaElement::validationMessage const):
(WebCore::HTMLTextAreaElement::valueMissing const):
(WebCore::HTMLTextAreaElement::tooShort const):
(WebCore::HTMLTextAreaElement::tooLong const):
* Source/WebCore/html/HTMLTextAreaElement.h:
* Source/WebCore/html/HTMLTextFormControlElement.cpp:
(WebCore::HTMLTextFormControlElement::insertedIntoAncestor):
(WebCore::HTMLTextFormControlElement::selectedText const):
(WebCore::HTMLTextFormControlElement::dispatchFormControlChangeEvent):
* Source/WebCore/html/HTMLTextFormControlElement.h:
(WebCore::HTMLTextFormControlElement::setTextAsOfLastFormControlChangeEvent):
* Source/WebCore/html/InputType.cpp:
(WebCore::InputType::fallbackValue const):
(WebCore::InputType::sanitizeValue const):
* Source/WebCore/html/InputType.h:
* Source/WebCore/html/MonthInputType.cpp:
(WebCore::MonthInputType::valueAsDate const):
* Source/WebCore/html/NumberInputType.cpp:
(WebCore::NumberInputType::valueAsDouble const):
(WebCore::NumberInputType::sanitizeValue const):
* Source/WebCore/html/NumberInputType.h:
* Source/WebCore/html/RadioNodeList.cpp:
(WebCore::nonEmptyRadioButton):
* Source/WebCore/html/RangeInputType.cpp:
(WebCore::RangeInputType::valueAsDouble const):
(WebCore::RangeInputType::setValue):
(WebCore::RangeInputType::fallbackValue const):
(WebCore::RangeInputType::sanitizeValue const):
* Source/WebCore/html/RangeInputType.h:
* Source/WebCore/html/SearchInputType.cpp:
(WebCore::SearchInputType::setValue):
* Source/WebCore/html/TextFieldInputType.cpp:
(WebCore::TextFieldInputType::setValue):
(WebCore::limitLength):
(WebCore::TextFieldInputType::sanitizeValue const):
* Source/WebCore/html/TextFieldInputType.h:
* Source/WebCore/html/URLInputType.cpp:
(WebCore::URLInputType::sanitizeValue const):
* Source/WebCore/html/URLInputType.h:

* Source/WebCore/html/parser/AtomHTMLToken.h:
(WebCore::AtomHTMLToken::initializeAttributes): Use raw pointers for unique-ing
because we&apos;re retaining them elsewhere and anyway we never dereference them.
Use a Vector instead of a hash table because number of elements is usually
small, so a Vector is faster.

* Source/WebCore/html/parser/HTMLDocumentParserFastPath.cpp:
(WebCore::HTMLFastPathParser::parseAttributes): Don&apos;t copy the empty string
unless you have to.
(WebCore::insertInUniquedSortedVector): Deleted.

* Source/WebCore/html/shadow/DateTimeFieldElement.cpp:
(WebCore::DateTimeFieldElement::visibleValue const):
* Source/WebCore/html/shadow/DateTimeFieldElement.h:
* Source/WebCore/html/shadow/DateTimeNumericFieldElement.cpp:
(WebCore::DateTimeNumericFieldElement::value const):
* Source/WebCore/html/shadow/DateTimeNumericFieldElement.h:
* Source/WebCore/html/shadow/DateTimeSymbolicFieldElement.cpp:
(WebCore::DateTimeSymbolicFieldElement::value const):
* Source/WebCore/html/shadow/DateTimeSymbolicFieldElement.h:
* Source/WebCore/html/shadow/SliderThumbElement.cpp:
(WebCore::sliderPosition):
* Source/WebCore/html/shadow/TextControlInnerElements.cpp:
(WebCore::SearchFieldCancelButtonElement::resolveCustomStyle):

* Source/WebCore/layout/formattingContexts/inline/InlineItemsBuilder.cpp:
(WebCore::Layout::InlineItemsBuilder::buildInlineItemListForTextFromBreakingPositionsCache):
* Source/WebCore/layout/formattingContexts/inline/text/TextUtil.cpp:
(WebCore::Layout::TextUtil::width):
(WebCore::Layout::TextUtil::trailingWhitespaceWidth):
(WebCore::Layout::TextUtil::fallbackFontsForText):
(WebCore::Layout::TextUtil::breakWord):
(WebCore::Layout::TextUtil::firstUserPerceivedCharacterLength):
* Source/WebCore/page/InteractionRegion.cpp:
(WebCore::shouldAllowNonInteractiveCursorForElement):
* Source/WebCore/platform/graphics/WidthIterator.cpp:
(WebCore::WidthIterator::applyFontTransforms):
(WebCore::WidthIterator::advanceInternal):
(WebCore::WidthIterator::applyExtraSpacingAfterShaping):
* Source/WebCore/rendering/RenderSearchField.cpp:
(WebCore::RenderSearchField::visibilityForCancelButton const):
* Source/WebCore/style/StyleAdjuster.cpp:
(WebCore::Style::Adjuster::adjustForSiteSpecificQuirks const):
(WebCore::Style::Adjuster::adjustmentForTextAutosizing): Don&apos;t copy for no reason.
Conversely, don&apos;t declare const auto&amp; when it has no meaning.

* Source/WebCore/rendering/style/RenderStyle.cpp:
(WebCore::RenderStyle::mutableFontDescriptionWithoutUpdate):
(WebCore::RenderStyle::mutableFontCascadeWithoutUpdate):
* Source/WebCore/rendering/style/RenderStyle.h: Helper functions for the setters
below.

* Source/WebCore/style/StyleAdjuster.cpp:
(WebCore::Style::Adjuster::adjustForSiteSpecificQuirks const):
(WebCore::Style::Adjuster::adjustmentForTextAutosizing):
* Source/WebCore/style/StyleBuilder.cpp:
(WebCore::Style::Builder::applyCustomPropertyImpl): Avoid refcount churn.

* Source/WebCore/style/StyleBuilderCustom.h:
(WebCore::Style::BuilderCustom::applyValueWebkitLocale):
(WebCore::Style::BuilderCustom::applyInitialFontFamily):
(WebCore::Style::BuilderCustom::applyInheritFontFamily):
(WebCore::Style::BuilderCustom::applyValueFontFamily):
(WebCore::Style::BuilderCustom::applyInheritFontVariantLigatures):
(WebCore::Style::BuilderCustom::applyInitialFontVariantLigatures):
(WebCore::Style::BuilderCustom::applyValueFontVariantLigatures):
(WebCore::Style::BuilderCustom::applyInheritFontVariantNumeric):
(WebCore::Style::BuilderCustom::applyInitialFontVariantNumeric):
(WebCore::Style::BuilderCustom::applyValueFontVariantNumeric):
(WebCore::Style::BuilderCustom::applyInheritFontVariantEastAsian):
(WebCore::Style::BuilderCustom::applyInitialFontVariantEastAsian):
(WebCore::Style::BuilderCustom::applyValueFontVariantEastAsian):
(WebCore::Style::BuilderCustom::applyInheritFontVariantAlternates):
(WebCore::Style::BuilderCustom::applyInitialFontVariantAlternates):
(WebCore::Style::BuilderCustom::applyInheritFontSize):
(WebCore::Style::BuilderCustom::applyValueFontSize):
(WebCore::Style::BuilderCustom::applyValueFontSizeAdjust):
* Source/WebCore/style/StyleBuilderState.cpp:
(WebCore::Style::BuilderState::updateFontForZoomChange):
* Source/WebCore/style/StyleBuilderState.h:
* Source/WebCore/style/StyleBuilderStateInlines.h:
(WebCore::Style::BuilderState::setFontDescription):
(WebCore::Style::BuilderState::setUsedZoom):
(WebCore::Style::BuilderState::setFontDescriptionKeywordSizeFromIdentifier):
(WebCore::Style::BuilderState::setFontDescriptionIsAbsoluteSize):
(WebCore::Style::BuilderState::setFontDescriptionFontSize):
(WebCore::Style::BuilderState::setFontDescriptionFamilies):
(WebCore::Style::BuilderState::setFontDescriptionIsSpecifiedFont):
(WebCore::Style::BuilderState::setFontDescriptionFeatureSettings):
(WebCore::Style::BuilderState::setFontDescriptionFontPalette):
(WebCore::Style::BuilderState::setFontDescriptionFontSizeAdjust):
(WebCore::Style::BuilderState::setFontDescriptionFontSmoothing):
(WebCore::Style::BuilderState::setFontDescriptionFontSynthesisSmallCaps):
(WebCore::Style::BuilderState::setFontDescriptionFontSynthesisStyle):
(WebCore::Style::BuilderState::setFontDescriptionFontSynthesisWeight):
(WebCore::Style::BuilderState::setFontDescriptionKerning):
(WebCore::Style::BuilderState::setFontDescriptionOpticalSizing):
(WebCore::Style::BuilderState::setFontDescriptionSpecifiedLocale):
(WebCore::Style::BuilderState::setFontDescriptionTextAutospace):
(WebCore::Style::BuilderState::setFontDescriptionTextRenderingMode):
(WebCore::Style::BuilderState::setFontDescriptionTextSpacingTrim):
(WebCore::Style::BuilderState::setFontDescriptionVariantCaps):
(WebCore::Style::BuilderState::setFontDescriptionVariantEmoji):
(WebCore::Style::BuilderState::setFontDescriptionVariantPosition):
(WebCore::Style::BuilderState::setFontDescriptionVariationSettings):
(WebCore::Style::BuilderState::setFontDescriptionWeight):
(WebCore::Style::BuilderState::setFontDescriptionWidth):
(WebCore::Style::BuilderState::setFontDescriptionVariantAlternates):
(WebCore::Style::BuilderState::setFontDescriptionVariantEastAsianVariant):
(WebCore::Style::BuilderState::setFontDescriptionVariantEastAsianWidth):
(WebCore::Style::BuilderState::setFontDescriptionVariantEastAsianRuby):
(WebCore::Style::BuilderState::setFontDescriptionKeywordSize):
(WebCore::Style::BuilderState::setFontDescriptionVariantCommonLigatures):
(WebCore::Style::BuilderState::setFontDescriptionVariantDiscretionaryLigatures):
(WebCore::Style::BuilderState::setFontDescriptionVariantHistoricalLigatures):
(WebCore::Style::BuilderState::setFontDescriptionVariantContextualAlternates):
(WebCore::Style::BuilderState::setFontDescriptionVariantNumericFigure):
(WebCore::Style::BuilderState::setFontDescriptionVariantNumericSpacing):
(WebCore::Style::BuilderState::setFontDescriptionVariantNumericFraction):
(WebCore::Style::BuilderState::setFontDescriptionVariantNumericOrdinal):
(WebCore::Style::BuilderState::setFontDescriptionVariantNumericSlashedZero): Added
efficient setters that avoid copying FontCascadeDescription. FontCascadeDescription
is 192 bytes, plus some out of line Vector storage, plus some String refcounts.
Copying a FontCascadeDescription just to change a bool is expensive. Copying
a FontCascadeDescription when nothing has changed is just funny.

* Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
(WebKit::WebPage::requestDocumentEditingContext):
* Source/WebKitLegacy/ios/WebCoreSupport/WebVisiblePosition.mm:
(-[DOMHTMLInputElement endPosition]):
(-[DOMHTMLTextAreaElement endPosition]):
* Source/WebKitLegacy/mac/DOM/DOMHTMLInputElement.mm:
(-[DOMHTMLInputElement value]):
* Source/WebKitLegacy/mac/DOM/DOMHTMLTextAreaElement.mm:
(-[DOMHTMLTextAreaElement value]):

Canonical link: <a href="https://commits.webkit.org/293816@main">https://commits.webkit.org/293816@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/75eb823f5780f7f84ff7b5545679544664572e88

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/100009 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/19657 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/9949 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/105137 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/50590 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/102050 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/19963 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/28130 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/76144 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/33218 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/103016 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/132/builds/15247 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/90324 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/56504 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/15056 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/8319 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/49959 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/92667 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/84978 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/8404 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/107497 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/98616 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/27122 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/19868 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/85102 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/27485 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/86529 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/84629 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/29293 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/7024 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/20958 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/16272 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/27059 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/32287 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/122242 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/26870 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/34126 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/30186 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/28429 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

kmiller68 commented on Source/JavaScriptCore/runtime/JSGlobalObjectFunctions.cpp:
```diff
@@ -1049,6 +1052,8 @@ JSC_DEFINE_HOST_FUNCTION(globalFuncCloneObject, (JSGlobalObject* globalObject, C
 
         target->putOwnDataPropertyBatching(vm, properties.data(), values.data(), properties.size());
+
+        ensureStillAliveHere(sourceStructure);
```
I think an `EnsureStillAliveScope` is a better idiom

kmiller68 commented on Source/JavaScriptCore/runtime/JSObject.h:
```diff
@@ -878,7 +878,7 @@ class JSObject : public JSCell {
     bool putOwnDataPropertyMayBeIndex(JSGlobalObject*, PropertyName, JSValue, PutPropertySlot&);
 
-    void putOwnDataPropertyBatching(VM&, const RefPtr<UniquedStringImpl>*, const EncodedJSValue*, unsigned size);
+    void putOwnDataPropertyBatching(VM&, UniquedStringImpl**, const EncodedJSValue*, unsigned size);
```
Should this be a `std::span<UniqueStringImpl*>`?

geoffreygaren replied:
It should be. But we currently skip enforcement in this file. I think I'll fix that in a follow-up, to ensure the integrity of the perf test.

kmiller68 commented on Source/JavaScriptCore/runtime/ObjectConstructorInlines.h:
```diff
@@ -337,6 +337,8 @@ ALWAYS_INLINE bool objectAssignFast(JSGlobalObject* globalObject, JSFinalObject*
     if (properties.size())
         target->putOwnDataPropertyBatching(vm, properties.data(), values.data(), properties.size());
+
+    ensureStillAliveHere(sourceStructure);
```
Ditto on `EnsureStillAliveScope`

kmiller68 commented on Source/JavaScriptCore/runtime/JSGlobalObjectFunctions.cpp:
```diff
@@ -962,8 +963,10 @@ JSC_DEFINE_HOST_FUNCTION(globalFuncCopyDataProperties, (JSGlobalObject* globalOb
+                target->putDirect(vm, properties[i], values.at(i));
         }
+
+        ensureStillAliveHere(sourceStructure);
```
Ditto.

kmiller68 reviewed (commented):
Some comments.

kmiller68 reviewed (approved):
r=me on JSC/WTF changes.

cdumez reviewed (approved):
WebCore & WebKit LGTM

Constellation reviewed (approved):
r=me if JetStream2 / JetStream3 are neutral.

kmiller68 commented on Source/JavaScriptCore/bytecode/DirectEvalCodeCache.h:
```diff
@@ -39,11 +39,20 @@ namespace JSC {
 
     class DirectEvalCodeCache {
     public:
+        enum class RopeSuffix : uint8_t {
+            None,
+            FunctionCall
+        };
```
Oof, this exact optimization is being removed in JS3 because we assumed it wasn't useful in the wild... Oh well I guess we can remove it in a year or so...

geoffreygaren replied:
I'm honestly not sure if people write code like this or not. I decided to err on the side of optimizing the code we've seen.

kmiller68 commented on Source/JavaScriptCore/heap/GCOwnedDataScope.h:
```diff
@@ -66,6 +67,13 @@ class JSCell;
 
 template<typename T>
 struct GCOwnedDataScope {
+    void* operator new(size_t) = delete;
```
Should be `WTF_FORBID_HEAP_ALLOCATION`

kmiller68 commented on Source/JavaScriptCore/heap/GCOwnedDataScope.h:
```diff
@@ -66,6 +67,13 @@ class JSCell;
+    GCOwnedDataScope()
+        : owner(nullptr)
+        , data(nullptr)
+    { }
```
We should just have default initializers for the members below:
```
const JSCell* owner { nullptr };
SUPPRESS_UNCOUNTED_MEMBER T data { nullptr };
```

geoffreygaren replied:
I do still need to offer `GCOwnedDataScope() = default;` after I make this change; otherwise the two argument constructor shadows the zero argument constructor. So I'll do it that way.

kmiller68 commented on Source/JavaScriptCore/runtime/JSStringJoiner.cpp:
```diff
@@ -165,6 +165,55 @@ static inline String joinStrings(const JSStringJoiner::Entries& strings, std::sp
+        return { };
+    }
+
+    switch (separator.size()) {
+    case 0: {
```
Nit: why a switch over an `if ... else`?

geoffreygaren replied:
Not a strong preference, but I was trying to mimic the structure of the more complex version of `joinStrings` above.

kmiller68 commented on Source/JavaScriptCore/runtime/JSStringJoiner.h:
```diff
@@ -217,4 +217,49 @@ ALWAYS_INLINE void JSStringJoiner::appendNumber(VM& vm, double value)
+            m_isAll8Bit &= string->is8Bit();
+        }
+
+        return joinSlow(globalObject, data, length);
```
Nit: Seems more like a `joinImpl` rather than `joinSlow` since you'll still call this in most cases.

kmiller68 reviewed (approved):
r=me on JSC changes with some comments

anttijk reviewed (approved):
I looked through the style related bits and they look good 🚢

geoffreygaren committed:
StringImpl refcounting should be atomic
https://bugs.webkit.org/show_bug.cgi?id=289962
rdar://147313174

Reviewed by Antti Koivisto, Yusuke Suzuki, Chris Dumez, and Keith Miller.

This patch takes the first step to making WTF::String thread-safe: Making
m_refcount increment and decrement atomic. (Hash, flags, and the AtomString
table are not yet atomic. I plan to address those in my next patch.)

The direct code change is pretty small, but it measured as a performance
regression. So most of this code is optimization. I got rid of about 1/3 of
WTF::StringImpl refcounts on Speedometer and about 1/2 on JetStream.

ARM has a special ldadd instruction for fast atomics. Net net, it measures as
about 1.5X slower than just add. That's pretty good! But it still required
some optimizations to be affordable. Intel atomics are much slower, so it took
far more optimizations to get Intel across the finish line.

A/B testers suggest that we might still have a < 0.5% JetStream regression on
Intel. I checked in with Yusuke, Keith, and Maciej, and we can move forward
because we have a large total JetStream speedup on Intel this year, so customers
on older hardware will still see a net speedup when they get new software.

* Source/JavaScriptCore/bytecode/DirectEvalCodeCache.cpp:
(JSC::DirectEvalCodeCache::setSlow):
* Source/JavaScriptCore/bytecode/DirectEvalCodeCache.h:
(JSC::DirectEvalCodeCache::CacheKey::CacheKey):
(JSC::DirectEvalCodeCache::CacheKey::hash const):
(JSC::DirectEvalCodeCache::CacheKey::operator== const):
(JSC::DirectEvalCodeCache::CacheLookupKey::CacheLookupKey):
(JSC::DirectEvalCodeCache::CacheLookupKey::hash const):
(JSC::DirectEvalCodeCache::CacheLookupKey::operator== const):
(JSC::DirectEvalCodeCache::CacheLookupKey::operator CacheKey const):
(JSC::DirectEvalCodeCache::CacheLookupKeyHashTranslator::hash):
(JSC::DirectEvalCodeCache::CacheLookupKeyHashTranslator::equal):
(JSC::DirectEvalCodeCache::get):
(JSC::DirectEvalCodeCache::set):
(JSC::DirectEvalCodeCache::tryGet): Deleted.
* Source/JavaScriptCore/interpreter/Interpreter.cpp:
(JSC::directEvalCacheKey):
(JSC::eval): Added a fast path for looking up a rope string without resolving
the rope when the right hand side of the rope is trivial.

* Source/JavaScriptCore/dfg/DFGOperations.cpp:
(JSC::DFG::JSC_DEFINE_JIT_OPERATION): No need to ref StringImpls here because
the Structure they came from ensures their lifetime in this scope.

* Source/JavaScriptCore/jit/JITOperations.cpp:
(JSC::putByVal):
(JSC::putByValOptimize):
(JSC::getByVal):
(JSC::JSC_DEFINE_JIT_OPERATION):
(JSC::getByValWithThis):
(JSC::getByValMegamorphic): Refactored fast paths not to ref strings.

* Source/JavaScriptCore/jit/JITWorklist.cpp:
(JSC::JITWorklist::enqueue): Dan and I collided here on
https://commits.webkit.org/293052@main, but this version measured as a speedup
so I'm keeping the slight change in how we load this value in case it matters.

* Source/JavaScriptCore/parser/Parser.cpp:
(JSC::Parser<LexerType>::parseInner):
(JSC::Parser<LexerType>::parseModuleSourceElements):
* Source/JavaScriptCore/parser/Parser.h:
(JSC::Scope::hasDeclaredVariable):
(JSC::Scope::hasLexicallyDeclaredVariable const):
(JSC::Scope::hasDeclaredParameter):
* Source/JavaScriptCore/parser/VariableEnvironment.cpp:
(JSC::VariableEnvironment::markVariableAsCapturedIfDefined):
(JSC::VariableEnvironment::markVariableAsCaptured):
(JSC::VariableEnvironment::markVariableAsImported):
(JSC::VariableEnvironment::markVariableAsExported):
* Source/JavaScriptCore/parser/VariableEnvironment.h:
(JSC::VariableEnvironment::contains const):
(JSC::VariableEnvironment::remove):
(JSC::VariableEnvironment::find):
(JSC::VariableEnvironment::find const):
* Source/JavaScriptCore/runtime/ArgList.h:
(JSC::MarkedVector::begin):
(JSC::MarkedVector::end): Removed some unnecessary conversions from StringImpl*
to RefPtr<StringImpl> for lookup.

* Source/JavaScriptCore/runtime/ArrayPrototype.cpp:
(JSC::fastJoin):
(JSC::JSC_DEFINE_HOST_FUNCTION): Added a fast path for joins that are all
strings.

* Source/JavaScriptCore/runtime/CacheableIdentifier.h:
* Source/JavaScriptCore/runtime/CacheableIdentifierInlines.h: Added a fast path
for grabbing a cacheable identifier without first asking if it exists.

* Source/JavaScriptCore/runtime/JSGlobalObjectFunctions.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION): No need for refcount churn if our StringImpl*
is owned by our structure.

* Source/JavaScriptCore/runtime/JSONAtomStringCache.h:
(JSC::JSONAtomStringCache::makeIdentifier): Deleted.
* Source/JavaScriptCore/runtime/JSONAtomStringCacheInlines.h:
(JSC::JSONAtomStringCache::makeIdentifier):
(JSC::JSONAtomStringCache::existingIdentifier):
(JSC::JSONAtomStringCache::make): Deleted. Added a fast path for grabbing a
StringImpl* without refcount churn.

* Source/JavaScriptCore/runtime/JSObject.cpp:
(JSC::JSObject::putOwnDataPropertyBatching):
(JSC::JSObject::putDirectForJSONSlow): Deleted.
* Source/JavaScriptCore/runtime/JSObject.h:
* Source/JavaScriptCore/runtime/JSObjectInlines.h:
(JSC::JSObject::putDirectInternal): Our fast path doesn't deterministically rule
out a hit anymore, so we can't skip that check on the slow path anymore.

* Source/JavaScriptCore/runtime/JSString.cpp:
(JSC:: const):
(JSC::JSRopeString::resolveRopeToAtomString const): Deleted.
(JSC::JSRopeString::resolveRopeToExistingAtomString const): Deleted.
* Source/JavaScriptCore/runtime/JSString.h:
(JSC::JSString::getValueImpl const):
(JSC::JSString::tryGetValueImpl const):
(JSC:: const):
(JSC::JSString::toAtomString const): Deleted.
(JSC::JSString::toExistingAtomString const): Deleted.
* Source/JavaScriptCore/runtime/JSStringInlines.h:
(JSC::JSRopeString::tryGetLHS const): Use GCOwnedDataScope for more return values, to avoid
refcount churn.

* Source/JavaScriptCore/runtime/JSStringJoiner.cpp:
(JSC::joinStrings):
(JSC::JSOnlyStringsJoiner::joinSlow):
* Source/JavaScriptCore/runtime/JSStringJoiner.h:
(JSC::JSOnlyStringsJoiner::JSOnlyStringsJoiner):
(JSC::JSOnlyStringsJoiner::tryJoin): Helper class for the fast path for joining
only strings.

* Source/JavaScriptCore/runtime/LiteralParser.cpp:
(JSC::reviverMode>::equalIdentifier):
(JSC::reviverMode>::existingIdentifier):
(JSC::reviverMode>::makeJSString):
(JSC::requires):
* Source/JavaScriptCore/runtime/LiteralParser.h: Added a fast path for following
a property transition without looking up and/or ref'ing an Identifier.

* Source/JavaScriptCore/runtime/NumericStrings.h: Made the cache bigger because
it made a ~20% difference on hit rate, reducing string churn.

* Source/JavaScriptCore/runtime/ObjectConstructor.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/JavaScriptCore/runtime/ObjectConstructorInlines.h:
(JSC::objectAssignFast): No need to ref strings owned by our structure.

* Source/JavaScriptCore/runtime/StringPrototype.cpp:
(JSC::substituteBackreferences):
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/JavaScriptCore/runtime/StringPrototypeInlines.h:
(JSC::jsSpliceSubstringsWithSeparator):
(JSC::jsSpliceSubstrings):
(JSC::removeAllUsingRegExpSearch):
(JSC::addToRegExpSearchCache):
(JSC::replaceAllWithCacheUsingRegExpSearchThreeArguments):
(JSC::replaceAllWithCacheUsingRegExpSearch):
(JSC::replaceAllWithStringUsingRegExpSearchNoBackreferences):
(JSC::replaceAllWithStringUsingRegExpSearch):
(JSC::replaceOneWithStringUsingRegExpSearch):
(JSC::replaceUsingRegExpSearch):
(JSC::removeUsingRegExpSearch): Deleted.
(JSC::replaceUsingRegExpSearchWithCache): Deleted.
* Source/JavaScriptCore/runtime/Structure.cpp:
(JSC::StructureTransitionTable::add): Streamlined the fast paths and added
a new fast path for a single separator join.

* Source/WTF/wtf/ValueOrReference.h: Made a new type to encapsulate a pattern
where we usually return an existing reference, but sometimes need to allocate
a temporary. This avoids a copy in the case where we return an existing reference.
Mostly deployed this in form controls code.

* Source/WTF/wtf/Vector.h:
(WTF::insertInUniquedSortedVector): Factored out a helper function for use in
HTML attributes.

* Source/WTF/wtf/text/StringImpl.cpp:
* Source/WTF/wtf/text/StringImpl.h:
(WTF::StringImpl::ref):
(WTF::StringImpl::deref): Aha! The actual refcounting change. (Note that memory order
is relaxed because we have no data dependencies yet. That will come next.)

* Source/WebCore/dom/Attribute.h:
(WebCore::Attribute::Attribute): Don't copy strings when moving.

* Source/WebCore/dom/CharacterData.h: Don't copy the empty string unless you're
actually going to use it.

* Source/WebCore/dom/ElementTextDirection.cpp:
(WebCore::computeAutoDirectionality):
* Source/WebCore/dom/Text.cpp:
(WebCore::Text::create): Deleted.
* Source/WebCore/dom/Text.h:
(WebCore::Text::create):
* Source/WebCore/editing/cocoa/AutofillElements.cpp:
(WebCore::AutofillElements::computeAutofillElements):
* Source/WebCore/editing/cocoa/HTMLConverter.mm:
(HTMLConverter::_processElement):
* Source/WebCore/html/BaseCheckableInputType.cpp:
(WebCore::BaseCheckableInputType::fallbackValue const):
* Source/WebCore/html/BaseCheckableInputType.h:
* Source/WebCore/html/BaseDateAndTimeInputType.cpp:
(WebCore::BaseDateAndTimeInputType::hasBadInput const):
(WebCore::BaseDateAndTimeInputType::sanitizeValue const):
(WebCore::BaseDateAndTimeInputType::updateInnerTextValue):
(WebCore::BaseDateAndTimeInputType::setupDateTimeChooserParameters):
* Source/WebCore/html/BaseDateAndTimeInputType.h:
* Source/WebCore/html/ColorInputType.cpp:
(WebCore::ColorInputType::fallbackValue const):
(WebCore::ColorInputType::sanitizeValue const):
(WebCore::ColorInputType::valueAsColor const):
* Source/WebCore/html/ColorInputType.h:
* Source/WebCore/html/DateTimeLocalInputType.cpp:
(WebCore::DateTimeLocalInputType::sanitizeValue const):
* Source/WebCore/html/DateTimeLocalInputType.h:
* Source/WebCore/html/EmailInputType.cpp:
(WebCore::EmailInputType::sanitizeValue const):
* Source/WebCore/html/EmailInputType.h:
* Source/WebCore/html/HTMLInputElement.cpp:
(WebCore::HTMLInputElement::tooShort const):
(WebCore::HTMLInputElement::tooLong const):
(WebCore::HTMLInputElement::computeValidity const):
(WebCore::HTMLInputElement::listOptionValueAsDouble):
(WebCore::HTMLInputElement::value const):
(WebCore::HTMLInputElement::valueWithDefault const):
(WebCore::HTMLInputElement::setValue):
(WebCore::HTMLInputElement::sanitizeValue const):
* Source/WebCore/html/HTMLInputElement.h:
* Source/WebCore/html/HTMLTextAreaElement.cpp:
(WebCore::HTMLTextAreaElement::childrenChanged):
(WebCore::HTMLTextAreaElement::appendFormData):
(WebCore::HTMLTextAreaElement::value const):
(WebCore::HTMLTextAreaElement::setValueCommon):
(WebCore::HTMLTextAreaElement::validationMessage const):
(WebCore::HTMLTextAreaElement::valueMissing const):
(WebCore::HTMLTextAreaElement::tooShort const):
(WebCore::HTMLTextAreaElement::tooLong const):
* Source/WebCore/html/HTMLTextAreaElement.h:
* Source/WebCore/html/HTMLTextFormControlElement.cpp:
(WebCore::HTMLTextFormControlElement::insertedIntoAncestor):
(WebCore::HTMLTextFormControlElement::selectedText const):
(WebCore::HTMLTextFormControlElement::dispatchFormControlChangeEvent):
* Source/WebCore/html/HTMLTextFormControlElement.h:
(WebCore::HTMLTextFormControlElement::setTextAsOfLastFormControlChangeEvent):
* Source/WebCore/html/InputType.cpp:
(WebCore::InputType::fallbackValue const):
(WebCore::InputType::sanitizeValue const):
* Source/WebCore/html/InputType.h:
* Source/WebCore/html/MonthInputType.cpp:
(WebCore::MonthInputType::valueAsDate const):
* Source/WebCore/html/NumberInputType.cpp:
(WebCore::NumberInputType::valueAsDouble const):
(WebCore::NumberInputType::sanitizeValue const):
* Source/WebCore/html/NumberInputType.h:
* Source/WebCore/html/RadioNodeList.cpp:
(WebCore::nonEmptyRadioButton):
* Source/WebCore/html/RangeInputType.cpp:
(WebCore::RangeInputType::valueAsDouble const):
(WebCore::RangeInputType::setValue):
(WebCore::RangeInputType::fallbackValue const):
(WebCore::RangeInputType::sanitizeValue const):
* Source/WebCore/html/RangeInputType.h:
* Source/WebCore/html/SearchInputType.cpp:
(WebCore::SearchInputType::setValue):
* Source/WebCore/html/TextFieldInputType.cpp:
(WebCore::TextFieldInputType::setValue):
(WebCore::limitLength):
(WebCore::TextFieldInputType::sanitizeValue const):
* Source/WebCore/html/TextFieldInputType.h:
* Source/WebCore/html/URLInputType.cpp:
(WebCore::URLInputType::sanitizeValue const):
* Source/WebCore/html/URLInputType.h:

* Source/WebCore/html/parser/AtomHTMLToken.h:
(WebCore::AtomHTMLToken::initializeAttributes): Use raw pointers for unique-ing
because we're retaining them elsewhere and anyway we never dereference them.
Use a Vector instead of a hash table because number of elements is usually
small, so a Vector is faster.

* Source/WebCore/html/parser/HTMLDocumentParserFastPath.cpp:
(WebCore::HTMLFastPathParser::parseAttributes): Don't copy the empty string
unless you have to.
(WebCore::insertInUniquedSortedVector): Deleted.

* Source/WebCore/html/shadow/DateTimeFieldElement.cpp:
(WebCore::DateTimeFieldElement::visibleValue const):
* Source/WebCore/html/shadow/DateTimeFieldElement.h:
* Source/WebCore/html/shadow/DateTimeNumericFieldElement.cpp:
(WebCore::DateTimeNumericFieldElement::value const):
* Source/WebCore/html/shadow/DateTimeNumericFieldElement.h:
* Source/WebCore/html/shadow/DateTimeSymbolicFieldElement.cpp:
(WebCore::DateTimeSymbolicFieldElement::value const):
* Source/WebCore/html/shadow/DateTimeSymbolicFieldElement.h:
* Source/WebCore/html/shadow/SliderThumbElement.cpp:
(WebCore::sliderPosition):
* Source/WebCore/html/shadow/TextControlInnerElements.cpp:
(WebCore::SearchFieldCancelButtonElement::resolveCustomStyle):

* Source/WebCore/layout/formattingContexts/inline/InlineItemsBuilder.cpp:
(WebCore::Layout::InlineItemsBuilder::buildInlineItemListForTextFromBreakingPositionsCache):
* Source/WebCore/layout/formattingContexts/inline/text/TextUtil.cpp:
(WebCore::Layout::TextUtil::width):
(WebCore::Layout::TextUtil::trailingWhitespaceWidth):
(WebCore::Layout::TextUtil::fallbackFontsForText):
(WebCore::Layout::TextUtil::breakWord):
(WebCore::Layout::TextUtil::firstUserPerceivedCharacterLength):
* Source/WebCore/page/InteractionRegion.cpp:
(WebCore::shouldAllowNonInteractiveCursorForElement):
* Source/WebCore/platform/graphics/WidthIterator.cpp:
(WebCore::WidthIterator::applyFontTransforms):
(WebCore::WidthIterator::advanceInternal):
(WebCore::WidthIterator::applyExtraSpacingAfterShaping):
* Source/WebCore/rendering/RenderSearchField.cpp:
(WebCore::RenderSearchField::visibilityForCancelButton const):
* Source/WebCore/style/StyleAdjuster.cpp:
(WebCore::Style::Adjuster::adjustForSiteSpecificQuirks const):
(WebCore::Style::Adjuster::adjustmentForTextAutosizing): Don't copy for no reason.
Conversely, don't declare const auto& when it has no meaning.

* Source/WebCore/rendering/style/RenderStyle.cpp:
(WebCore::RenderStyle::mutableFontDescriptionWithoutUpdate):
(WebCore::RenderStyle::mutableFontCascadeWithoutUpdate):
* Source/WebCore/rendering/style/RenderStyle.h: Helper functions for the setters
below.

* Source/WebCore/style/StyleAdjuster.cpp:
(WebCore::Style::Adjuster::adjustForSiteSpecificQuirks const):
(WebCore::Style::Adjuster::adjustmentForTextAutosizing):
* Source/WebCore/style/StyleBuilder.cpp:
(WebCore::Style::Builder::applyCustomPropertyImpl): Avoid refcount churn.

* Source/WebCore/style/StyleBuilderCustom.h:
(WebCore::Style::BuilderCustom::applyValueWebkitLocale):
(WebCore::Style::BuilderCustom::applyInitialFontFamily):
(WebCore::Style::BuilderCustom::applyInheritFontFamily):
(WebCore::Style::BuilderCustom::applyValueFontFamily):
(WebCore::Style::BuilderCustom::applyInheritFontVariantLigatures):
(WebCore::Style::BuilderCustom::applyInitialFontVariantLigatures):
(WebCore::Style::BuilderCustom::applyValueFontVariantLigatures):
(WebCore::Style::BuilderCustom::applyInheritFontVariantNumeric):
(WebCore::Style::BuilderCustom::applyInitialFontVariantNumeric):
(WebCore::Style::BuilderCustom::applyValueFontVariantNumeric):
(WebCore::Style::BuilderCustom::applyInheritFontVariantEastAsian):
(WebCore::Style::BuilderCustom::applyInitialFontVariantEastAsian):
(WebCore::Style::BuilderCustom::applyValueFontVariantEastAsian):
(WebCore::Style::BuilderCustom::applyInheritFontVariantAlternates):
(WebCore::Style::BuilderCustom::applyInitialFontVariantAlternates):
(WebCore::Style::BuilderCustom::applyInheritFontSize):
(WebCore::Style::BuilderCustom::applyValueFontSize):
(WebCore::Style::BuilderCustom::applyValueFontSizeAdjust):
* Source/WebCore/style/StyleBuilderState.cpp:
(WebCore::Style::BuilderState::updateFontForZoomChange):
* Source/WebCore/style/StyleBuilderState.h:
* Source/WebCore/style/StyleBuilderStateInlines.h:
(WebCore::Style::BuilderState::setFontDescription):
(WebCore::Style::BuilderState::setUsedZoom):
(WebCore::Style::BuilderState::setFontDescriptionKeywordSizeFromIdentifier):
(WebCore::Style::BuilderState::setFontDescriptionIsAbsoluteSize):
(WebCore::Style::BuilderState::setFontDescriptionFontSize):
(WebCore::Style::BuilderState::setFontDescriptionFamilies):
(WebCore::Style::BuilderState::setFontDescriptionIsSpecifiedFont):
(WebCore::Style::BuilderState::setFontDescriptionFeatureSettings):
(WebCore::Style::BuilderState::setFontDescriptionFontPalette):
(WebCore::Style::BuilderState::setFontDescriptionFontSizeAdjust):
(WebCore::Style::BuilderState::setFontDescriptionFontSmoothing):
(WebCore::Style::BuilderState::setFontDescriptionFontSynthesisSmallCaps):
(WebCore::Style::BuilderState::setFontDescriptionFontSynthesisStyle):
(WebCore::Style::BuilderState::setFontDescriptionFontSynthesisWeight):
(WebCore::Style::BuilderState::setFontDescriptionKerning):
(WebCore::Style::BuilderState::setFontDescriptionOpticalSizing):
(WebCore::Style::BuilderState::setFontDescriptionSpecifiedLocale):
(WebCore::Style::BuilderState::setFontDescriptionTextAutospace):
(WebCore::Style::BuilderState::setFontDescriptionTextRenderingMode):
(WebCore::Style::BuilderState::setFontDescriptionTextSpacingTrim):
(WebCore::Style::BuilderState::setFontDescriptionVariantCaps):
(WebCore::Style::BuilderState::setFontDescriptionVariantEmoji):
(WebCore::Style::BuilderState::setFontDescriptionVariantPosition):
(WebCore::Style::BuilderState::setFontDescriptionVariationSettings):
(WebCore::Style::BuilderState::setFontDescriptionWeight):
(WebCore::Style::BuilderState::setFontDescriptionWidth):
(WebCore::Style::BuilderState::setFontDescriptionVariantAlternates):
(WebCore::Style::BuilderState::setFontDescriptionVariantEastAsianVariant):
(WebCore::Style::BuilderState::setFontDescriptionVariantEastAsianWidth):
(WebCore::Style::BuilderState::setFontDescriptionVariantEastAsianRuby):
(WebCore::Style::BuilderState::setFontDescriptionKeywordSize):
(WebCore::Style::BuilderState::setFontDescriptionVariantCommonLigatures):
(WebCore::Style::BuilderState::setFontDescriptionVariantDiscretionaryLigatures):
(WebCore::Style::BuilderState::setFontDescriptionVariantHistoricalLigatures):
(WebCore::Style::BuilderState::setFontDescriptionVariantContextualAlternates):
(WebCore::Style::BuilderState::setFontDescriptionVariantNumericFigure):
(WebCore::Style::BuilderState::setFontDescriptionVariantNumericSpacing):
(WebCore::Style::BuilderState::setFontDescriptionVariantNumericFraction):
(WebCore::Style::BuilderState::setFontDescriptionVariantNumericOrdinal):
(WebCore::Style::BuilderState::setFontDescriptionVariantNumericSlashedZero): Added
efficient setters that avoid copying FontCascadeDescription. FontCascadeDescription
is 192 bytes, plus some out of line Vector storage, plus some String refcounts.
Copying a FontCascadeDescription just to change a bool is expensive. Copying
a FontCascadeDescription when nothing has changed is just funny.

* Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
(WebKit::WebPage::requestDocumentEditingContext):
* Source/WebKitLegacy/ios/WebCoreSupport/WebVisiblePosition.mm:
(-[DOMHTMLInputElement endPosition]):
(-[DOMHTMLTextAreaElement endPosition]):
* Source/WebKitLegacy/mac/DOM/DOMHTMLInputElement.mm:
(-[DOMHTMLInputElement value]):
* Source/WebKitLegacy/mac/DOM/DOMHTMLTextAreaElement.mm:
(-[DOMHTMLTextAreaElement value]):

Canonical link: https://commits.webkit.org/293816@main
